src/java.base/share/classes/sun/nio/fs/AbstractWatchKey.java

```diff
@@ -28,16 +28,30 @@
 import java.nio.file.*;
 import java.util.*;
 
+import sun.security.action.GetPropertyAction;
+
 /**
  * Base implementation class for watch keys.
  */
 
 abstract class AbstractWatchKey implements WatchKey {
 
     /**
-     * Maximum size of event list (in the future this may be tunable)
+     * Maximum size of event list
      */
-    static final int MAX_EVENT_LIST_SIZE    = 512;
+    @SuppressWarnings("removal")
+    static final int MAX_EVENT_LIST_SIZE;
+    static {
+        String rawValue = GetPropertyAction.privilegedGetProperty(
+            "jdk.nio.file.maxWatchEvents", "512");
+        int intValue;
+        try {
+            intValue = Integer.decode(rawValue);
+        } catch (NumberFormatException e) {
+            intValue = 512;
+        }
+        MAX_EVENT_LIST_SIZE = intValue;
```

Comment:
What if `intValue` turns on to be non-positive for some reason? Should the value also be clamped to some maximum positive value, perhaps determined by some characteristic of the system on which the code is running?

Reply:
If you are going the system property route, then something like this (not tested) might be in order:

```
    private static final int SYSTEM_MAX_EVENT_LIST_SIZE = 100_000; // value TBD
    private static final int DEFAULT_MAX_EVENT_LIST_SIZE = 512;

    static final int MAX_EVENT_LIST_SIZE;
    static {
        String rawValue = GetPropertyAction.privilegedGetProperty
            ("jdk.nio.file.maxWatchEvents",
            String.valueOf(DEFAULT_MAX_EVENT_LIST_SIZE));
        int intValue;
        try {
            intValue = Math.min(Math.max(Math.toIntExact(Long.decode(rawValue)),
                                         DEFAULT_MAX_EVENT_LIST_SIZE),
                                SYSTEM_MAX_EVENT_LIST_SIZE);
        } catch (NumberFormatException nfe) {
            intValue = DEFAULT_MAX_EVENT_LIST_SIZE;
        } catch (ArithmeticException ae) { // rawValue overflowed an int
            intValue = SYSTEM_MAX_EVENT_LIST_SIZE;
        }
        MAX_EVENT_LIST_SIZE = intValue;
    }
```

Reply:
Thanks for the suggestion, the validation was definitely lacking. I pushed a new version that uses `Math.clamp` to clamp to [512, Integer.MAX_VALUE] and thus at least cover the lower bound. 

I learned that e.g. NTFS allows up to 2^32 entries per directory (in theory, of course), so I don't see a strong argument in favor of an upper bound motivated by hardware. Do you have a particular bottleneck or problematic use case in mind?

What about the case where the event size reaches `Integer.MAX_SIZE`? An OVERFLOW event could be better than an OOM due to the list size getting too large, what do you think?

Reply:
I wrote a prototype of `WatchService` for macOS that used `kqueue(2)` and it ultimately was inviable because it used up all the file descriptors. That was what prompted my comment, but it is irrelevant, really.

In the absence of such a real world system constraint, possibly some moderately large upper bound could be established based on empirical observation. It you are reacting to the OVERFLOW events already, I doubt we need to get to `Integer.MAX_VALUE`. Maybe all that's needed is something large enough to reduce the frequency of OVERFLOW events. I don't have any insight into that.

Reply:
> It you are reacting to the OVERFLOW events already, I doubt we need to get to Integer.MAX_VALUE. Maybe all that's needed is something large enough to reduce the frequency of OVERFLOW events. I don't have any insight into that.

I searched for usage reports and found users trying to watch "billions" of files at a time. Given that the actual limit is still user-configurable and the default remains at 512, I'm not sure whether there is any reasonable hard limit I could come up with that doesn't unnecessarily restrict users. After all, *if* a user sets this limit so high that they run into e.g. memory issues due to a large event list, wouldn't we consider them responsible for this?

In lieu of an actual upper bound I could justify, I went with `ArraySupport.SOFT_MAX_ARRAY_LENGTH` for now so that an OVERFLOW event is likely to. be sent before the list OOMs when trying to grow.

Reply:
The proposed change looks overly complicated. If the property is property is set to garbage or a numeric value < 1 then there may be an argument to use the default but it seems very strange to disallow setting it to a value such as 128.

Reply:
I'm fully open to changing it so that the default is only used if `intValue <= 0`. 

An argument I could see in favor of enforcing a minimum of 512 is backwards compatibility: It's possible that there are applications out there relying on being able to receive at least 512 file system events without an (automatic) OVERFLOW. If users of those applications set a lower limit, this could result in subtle thrashing issues that may not be directly attributable to this setting. 

But this is just theory and I'm not sure what the right balance between this concern and simplicity is. @AlanBateman Curious to hear your thoughts on this.

Reply:
> An argument I could see in favor of enforcing a minimum of 512 is backwards compatibility

That was my thinking on this topic, but I would be okay with a minimum of 1.